pkg/state/address_transactions.go

```diff
@@ -297,97 +299,100 @@ func (at *addressTransactions) shouldPersist() (bool, error) {
+	size := fst.Size()
+	slog.Info("Starting to sort TransactionsByAddresses file, will take awhile...", "size", size)
 	debug.FreeOSMemory()
-	// Create file for emsort and set emsort over it.
-	tempFile, err := os.CreateTemp(os.TempDir(), "emsort")
-	if err != nil {
-		return errors.Wrap(err, "failed to create temp file for emsort")
-	}
-	defer func(name string) {
-		if rmErr := os.Remove(name); rmErr != nil {
-			slog.Warn("Failed to remove temporary file", logging.Error(rmErr))
-		}
-	}(tempFile.Name())
-	sort, err := emsort.NewFixedSize(addrTxRecordSize, maxEmsortMem, tempFile)
-	if err != nil {
-		return errors.Wrap(err, "emsort.NewFixedSize() failed")
-	}
-
-	// Read records from file and append to emsort.
-	for readPos := int64(0); readPos < size; readPos += addrTxRecordSize {
-		record := make([]byte, addrTxRecordSize)
-		if n, err := at.addrTransactions.ReadAt(record, readPos); err != nil {
-			return err
-		} else if n != addrTxRecordSize {
-			return errors.New("failed to read full record")
-		}
-		// Filtering optimization: if all blocks are valid,
-		// we shouldn't check isValid() on records.
-		isValid := true
-		if at.amend {
-			blockNum := binary.BigEndian.Uint32(record[proto.AddressIDSize : proto.AddressIDSize+4])
-			isValid, err = at.stateDB.isValidBlock(blockNum)
-			if err != nil {
-				return errors.Wrap(err, "isValidBlock() failed")
+	eg, egCtx := errgroup.WithContext(at.ctx)
+
+	conf := extsort.DefaultConfig()
+	conf.TempFilesDir = os.TempDir() // Set dir explicitly to turn off automatic selection.
+
+	inCh := make(chan []byte, conf.SortedChanBuffSize) // Set size of input channel the same as of output.
+	eg.Go(func() error {
+		defer close(inCh)
+		for readPos := int64(0); readPos < size; readPos += addrTxRecordSize {
+			rec := make([]byte, addrTxRecordSize)
+			if n, rErr := at.addrTransactions.ReadAt(rec, readPos); rErr != nil {
+				return rErr
+			} else if n != addrTxRecordSize {
+				return errors.New("failed to read full record")
 			}
+			// Filtering optimization: if all blocks are valid,
+			// we shouldn't check isValid() on records.
+			var (
+				isValid = true
+				vErr    error
+			)
+			if at.amend {
+				blockNum := binary.BigEndian.Uint32(rec[proto.AddressIDSize : proto.AddressIDSize+4])
+				isValid, vErr = at.stateDB.isValidBlock(blockNum)
+				if vErr != nil {
+					return fmt.Errorf("block validation failed: %w", vErr)
+				}
+			}
+			if !isValid {
+				// Invalid record, we should skip it.
+				continue
+			}
+			inCh <- rec
 		}
-		if !isValid {
-			// Invalid record, we should skip it.
-			continue
-		}
-		if err := sort.Push(record); err != nil {
-			return errors.Wrap(err, "emsort.Push() failed")
+		slog.Info("Finished to send transactions to sorter")
+		return nil
+	})
+	sorter, outCh, errCh := extsort.Generic(inCh, nopPassBytes, nopPassBytes, bytes.Compare, conf)
+	eg.Go(func() error {
+		sorter.Sort(egCtx)
+		return nil
+	})
+	eg.Go(func() error {
+		return <-errCh
```

Comment:
is it guaranteed that this `chan` will be closed? What about external context cancellation? Would `sorter.Sort(egCtx)` handle it?